File: solutions/java/0065_有效数字.java
public class Solution {
    public boolean isNumber(String s) {
        s = s.trim();
        if (s.isEmpty()) {
            return false;
        }
        boolean seenDigit = false;
        boolean seenDot = false;
        boolean seenExp = false;
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch == '+' || ch == '-') {
                if (i > 0 && s.charAt(i - 1) != 'e' && s.charAt(i - 1) != 'E') {
                    return false;
                }
            } else if (ch == '.') {
                if (seenDot || seenExp) {
                    return false;
                }
                seenDot = true;
            } else if (ch == 'e' || ch == 'E') {
                if (seenExp || !seenDigit) {
                    return false;
                }
                seenExp = true;
                seenDigit = false;
            } else if (Character.isDigit(ch)) {
                seenDigit = true;
            } else {
                return false;
            }
        }
        return seenDigit;
    }
}

File: solutions/go/0065_有效数字.go
func isNumber(s string) bool {
	s = strings.TrimSpace(s)
	if len(s) == 0 {
		return false
	}
	seenDigit := false
	seenDot := false
	seenExp := false
	for i, ch := range s {
		if ch == '+' || ch == '-' {
			if i > 0 && s[i-1] != 'e' && s[i-1] != 'E' {
				return false
			}
		} else if ch == '.' {
			if seenDot || seenExp {
				return false
			}
			seenDot = true
		} else if ch == 'e' || ch == 'E' {
			if seenExp || !seenDigit {
				return false
			}
			seenExp = true
			seenDigit = false
		} else if unicode.IsDigit(ch) {
			seenDigit = true
		} else {
			return false
		}
	}
	return seenDigit
}
